More flexibility with wildcards in selection

Iv-Hristov opened the pull request:
Fixes #2436

Changes made in this Pull Request:
 - Selection strings changed to use fnmatch. This now allows for more flexible wildcard usage as well as for using multiple wildcards at once. 
 - Added two new tests to match the new functionality. 


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

richardjgowers commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -379,6 +379,16 @@ def test_wildcard_terminal_selection(self, universe):
+        assert ag == ag_wild
+
+    def test_wildcard_double_selection(self, universe):
+        ag = universe.select_atoms('resname ASN or resname ASP or resname HSD')
```
a shortcut here is `resname ASN ASP HSD`

Iv-Hristov commented:
Thank you for the feedback. I implemented the suggested changes and added two more corner case tests that I could think of. As for the performance, I measured it using a bigger system (340K atoms) and the old version which doesn't support different wildcards performed slightly better (0.6s vs 0.9s). I am not aware of the underlying implementation of fnmatch but we could also try using the python re module.

richardjgowers commented on package/CHANGELOG:
```diff
@@ -15,7 +15,7 @@ The rules for this file:
 mm/dd/yy richardjgowers, kain88-de, lilyminium, p-j-smith, bdice, joaomcteixeira,
          PicoCentauri, davidercruz, jbarnoud, RMeli, IAlibay, mtiberti, CCook96,
-         Yuan-Yu, xiki-tempula
+         Yuan-Yu, xiki-tempula, Iv-Hristov
```
So because this is your first ever contribution, you also need to add your name to the `AUTHORS` file

richardjgowers commented on package/CHANGELOG:
```diff
@@ -56,6 +56,7 @@ Fixes
   * Added parmed to setup.py
 
 Enhancements
+  * Changed selection wildcards to support multiple wildcards
```
reference the initial issue #

richardjgowers commented on package/MDAnalysis/core/selection.py:
```diff
@@ -499,8 +500,7 @@ def apply(self, group):
 
-    Supports the use of one wildcard at the start, 
-    end, and middle of strings
+    Supports multiple wildcards, based on fnmatch
```
add a `.. versionchanged::` thing

richardjgowers commented on package/MDAnalysis/core/selection.py:
```diff
@@ -512,16 +512,8 @@ def __init__(self, parser, tokens):
-                mask |= np.char.startswith(values, val[:wc_pos])
-                mask &= np.char.endswith(values, val[wc_pos+1:])
-
+            values = getattr(group, self.field).astype(np.str_)
```
so I think I was calling `.astype(np.str_)` here because we were pushing it into a `np.char` function.  Now we're using `fnmatch` this likely isn't necessary, try removing this?

IAlibay commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -379,6 +379,25 @@ def test_wildcard_terminal_selection(self, universe):
         ag_wild = universe.select_atoms('resname AS*')
         assert ag == ag_wild
 
+    def test_wildcard_single_selection(self, universe):
```
Just a suggestion here, since these four tests have the same structure, you might want to squash them together using pytest.mark.parametrize.

IAlibay replied:
Sorry for the dealy in providing more details here. This is the sort of construct I would suggest going for:

```
@pytest.mark.parametrize('selstring, wildstring', [
    ('resname TYR THR', 'resname T?R'),
    ('resname ASN ASP HSD', 'resname *S?'),
    ('resname LEU LYS', 'resname L**'),
    ('resname MET', 'resname *M*')
])
def test_wildcard_selection(self, universe, selstring, wildstring):
    ag = universe.select_atoms(selstring)
    ag_wild = universe.select_atoms(wildstring)
    assert ag == ag_wild
```

Note: I've changed the `'resname T?R'` case to match https://github.com/MDAnalysis/mdanalysis/pull/2551#discussion_r384398177

Iv-Hristov commented:
@IAlibay @richardjgowers Thank you for the feedback! I have combined all the wildcard tests to eliminate conde duplication as Irfan suggested.

IAlibay commented on package/MDAnalysis/core/selection.py:
```diff
@@ -499,8 +500,8 @@ def apply(self, group):
-    Supports the use of one wildcard at the start, 
-    end, and middle of strings
+    .. versionchanged:: 0.21
+    Supports multiple wildcards, based on fnmatch
```
@Iv-Hristov the text entry in a `versionchanged` needs to be indented.

So:
```
.. versionchanged:: 1.0.0
   Supports...
```

Iv-Hristov commented:
@orbeckst Thank you for the feedback! I will make sure to do that later today.

IAlibay commented on package/doc/sphinx/source/documentation_pages/selections.rst:
```diff
@@ -104,6 +99,22 @@ moltype *molecule-type*
+    Is a pattern that will match multiple characters.  For example,
+    ``GL*`` selects all strings that start with "GL" such as "GLU",
+    "GLY", "GLX29", "GLN".
+``[seq]``
```
I believe that fnmatch will also support [!seq] , I assume this would also work here.

Either way, you should try to introduce a test for [seq] (and [!seq] if it works).

Iv-Hristov commented:
I think I documented all the functionality and added two more tests. However, I am having a problem where pytest.mark.parametrize complains if I try and use square brackets as an input because it expects a string. I tried a few escape characters such as '\', '\\', '\Q... \E' but nothing seemed to work. Does anyone have any experience with what escape sequence might work so that I can squish all the tests together?

This is what gives the error "tuple object not callable":

    @pytest.mark.parametrize('selstring, wildstring', [

        ('resname TYR THR', 'resname T*R'),
        ('resname ASN GLN', 'resname *N'),
        ('resname ASN ASP', 'resname AS*'),
        ('resname TYR THR', 'resname T?R'),
        ('resname ASN ASP HSD', 'resname *S?'),
        ('resname LEU LYS', 'resname L**'),
        ('resname MET', 'resname *M*')
        ('resname GLN GLU', 'resname GL[NY]')

    ])
    def test_wildcard_selection(self, universe, selstring, wildstring):
        ag = universe.select_atoms(selstring)
        ag_wild = universe.select_atoms(wildstring)
        assert ag == ag_wild
`

orbeckst commented:
You missed the comma after line `   ('resname MET', 'resname *M*')`.

orbeckst commented on package/doc/sphinx/source/documentation_pages/selections.rst:
```diff
@@ -104,6 +99,27 @@ moltype *molecule-type*
+Pattern matching
+----------------
+
+The pattern matching notation described bellow is used to specify 
```
below

orbeckst commented on package/doc/sphinx/source/documentation_pages/selections.rst:
```diff
@@ -104,6 +99,27 @@ moltype *molecule-type*
+----------------
+
+The pattern matching notation described bellow is used to specify 
+patterns for matching strings:
```
```suggestion
patterns for matching strings (based on :mod:`fnmatch`): 
```

orbeckst commented:
Congratulations @Iv-Hristov on your first merged PR. Nice contribution!
